### ISSUE TITLE
The requested version (1) is less than the existing version (3).

### ISSUE THREAD

**Issue by nielsnl68** [ujjwalguptaofficial/JsStore#218]
# Title 

How to work correctly with versioning

# Description 

I'm new with jsStore, and while developing my webapp i was happy to see that you implemented a versioning option.
But now i cant access my data anymore. I have the feeling i should set the version somewhere else as well, but reading your doc's i havent found it,

At the moment i have the version number only at the table definition, should i set it somewhere else as well?

**Comment by ujjwalguptaofficial** [comment]
could you provide me your db schema ?

**Comment by nielsnl68** [comment]
``` javascript
function getDbSchema() {
    var metingen = {
        name: 'metingen',
        version: 2,
        columns: {
            id:    { primaryKey: true, autoIncrement: true },
            meter: { notNull: true, dataType: 'string', default: 'Handmatig'},
            datum: { notNull: true,  dataType: 'string' },
            tijd: { notNull: true, dataType: 'string' },
            waarde: {notNull: true,  dataType: 'number' },
            maaltijd: { notNull: true, dataType: 'string', default: 'U' },
            importDT: { notNull: true, dataType: 'string' },
            unique: { keyPath: ['meter', 'datum','tijd'],unique:true }
        }
    };
    var gebruikers = {
        name: 'gebruikers',
        version: 3,
        columns: {
            id: { primaryKey: true, autoIncrement: true },
            naam: { notNull: true, dataType: 'string', unique: true },
            meter: { notNull: true, dataType: 'string', default: 'Handmatig' },
            model: { notNull: false, dataType: 'string', default: '' },
            importDT: { notNull: false, dataType: 'string' },
            exportDT: { notNull: false, dataType: 'string' },
            refID: { notNull: true, dataType: 'string', default: '' },
            pohID: { notNull: false, dataType: 'string', default: '' },
            key:   {  notNull: false, dataType: 'string', default: ''}
        }
    }

    var db = {
        name: 'metingen',
        tables: [gebruikers, metingen]
    }
    return db;
}
```

Somehow i managed to clear all browser app content. So i cant show what was on the browsers side. What i saw is that the version numbers where mentioned in the application section in the devtools.

**Comment by ujjwalguptaofficial** [comment]
There is a bug due to that you are getting issue. The solution is to simply add version in database and everything should works OK.

```
function getDbSchema() {
    var metingen = {
        name: 'metingen',
        version: 2,
        columns: {
            id:    { primaryKey: true, autoIncrement: true },
            meter: { notNull: true, dataType: 'string', default: 'Handmatig'},
            datum: { notNull: true,  dataType: 'string' },
            tijd: { notNull: true, dataType: 'string' },
            waarde: {notNull: true,  dataType: 'number' },
            maaltijd: { notNull: true, dataType: 'string', default: 'U' },
            importDT: { notNull: true, dataType: 'string' },
            unique: { keyPath: ['meter', 'datum','tijd'],unique:true }
        }
    };
    var gebruikers = {
        name: 'gebruikers',
        version: 3,
        columns: {
            id: { primaryKey: true, autoIncrement: true },
            naam: { notNull: true, dataType: 'string', unique: true },
            meter: { notNull: true, dataType: 'string', default: 'Handmatig' },
            model: { notNull: false, dataType: 'string', default: '' },
            importDT: { notNull: false, dataType: 'string' },
            exportDT: { notNull: false, dataType: 'string' },
            refID: { notNull: true, dataType: 'string', default: '' },
            pohID: { notNull: false, dataType: 'string', default: '' },
            key:   {  notNull: false, dataType: 'string', default: ''}
        }
    }

    var db = {
        name: 'metingen',
        tables: [gebruikers, metingen],
        version:3 //
    }
    return db;
}
```

**Comment by ujjwalguptaofficial** [comment]
@nielsnl68 issue is fixed in new version - 4.2.0. We have removed support for table version and added `upgrade` option. 

prevously we were calculating the upgrade value from table version but this has some edge cases like yours. So now taking `upgrade` value from users.

So if you want to preserve your data makes `upgrade` value false in table. For more info, please have a look at docs - https://jsstore.net/tutorial/update-schema/

**Comment by nielsnl68** [comment]
Okay thanks, a thing what i do not understand is why you (guys) decided to delete all data when you do an upgrade. In most of the databases you can (most of the time) do table schema changes without the lose of data. Only when the change is so drastic you need to help converting the data.
When there is no other solution, maybe you could introduce an event that can be   caught to do the load / upgrade / restore action.

En then there is that async function you shown on that page, i could not figure out how that should work.
You first do the select to `allData`  but where do we select the `{tableName}` variable? Is that something jsStore will fill in automatically ?

And then at the end your started the insert but you do not supply the `allData` variable. How does the call knows what to insert?

When it is this much code to do to preserve the database content, why not doing it automatically inside jsStore it self, when the engine see it needs to upgrade the schema?

**Comment by ujjwalguptaofficial** [comment]
no you need to provide tableName yourself and allData not being used was mistake in docs, i have fixed it. 

Regarding
> what i do not understand is why you (guys) decided to delete all data when you do an upgrade.

Frankly there were less documentation about whether we can change the existing table and that's why i have created like this. But now i found out - we can achieve this similar to other database.  I will work towards this.

Thanks for your question.

**Comment by ujjwalguptaofficial** [comment]
now you can alter the table without losing your data starting from v 4.2.1. Here is updated doc - https://jsstore.net/tutorial/update-schema/

I am glad that we had this conversation. Feel free to suggest any changes.

**Comment by nielsnl68** [comment]
Wauwsers, you are fast. 

the documention looks good. Just to be clear. What happens when an user start at version 2 from your example how does the shema looks like? Can the main columns stucture be as it is in version 2 and the alter {2: {}}  is only used for people that has a version 1 table? like:

``` javascript
var tblProduct = {
    name: 'Product',
    columns: {
        id: {
            primaryKey: true,
            autoIncrement: true,
            notNull: true
        },
        name: {
            dataType: 'string'
        }
    },
    alter: {
        // 2 is database version to target
        2: {
            modify: {
                id: {
                    notNull: true
                }
            },
            add: {
                name: {
                    dataType: 'string'
                }
            },
            drop: {
                count: {}
            },
            convert: function (tableName, oldVerson, newVersion) {
                // your suggestion 
                // change datatype from number to string
                var existingData = await select({
                    from: `${tableName}`
                });

                existingData = existingData.map(value => {
                    value.count = value.count.toString();
                    return value;
                })
                await insert({
                    into: `${tableName}`,
                    upsert: true,
                    values: [existingData]
                })

                // Or as my suggestion in one go:
                await update({
                    in: `{ tableName }`,
                    convert: (value) => {
                        value.count = value.count.toString();
                        return value;
                    }
                });
            }
        }
    }
}
var db = {
    name: "db_name",
    tables: [tblProduct],
    version: 2 //Default version is 1.
}

``` 
So the main column structure looks always like the most current version and the alter section is used only for older version?

that convertion section ... is that possible to put that also inside the alter structure.
as a function call. as shown above?

**Comment by ujjwalguptaofficial** [comment]
version in alter command is the version in your database, so when database is created with version 2. JsStore search for version 2 in alter command.

regarding convert function - no its not possible to pass the pass a function to web worker.

Although i can write the code to convert into new datatype automatically but there are so many cases where automation won't be a good solution.

e.g - consider you were saving a date time value as `string`, and now you altered the data type to `date_time` in new version. When being converted to datatype, there might be that there is no value in the column or value is not date convertable. 

That's why i decided that user should convert the value manually as it provides command over your code & data.

**Comment by nielsnl68** [comment]
> version in alter command is the version in your database, so when database is created with version 2. JsStore search for version 2 in alter command.
> 
not sure about what you mean here, sorry.  I tried to follow the jsStore code but i got lost.

> regarding convert function - no its not possible to pass the pass a function to web worker.
>
Okay then that out off the picture. 
  
> Although i can write the code to convert into new datatype automatically but there are so many cases where automation won't be a good solution.
>
No, no. I like to have control over the conversions. It is just how and when.

> 
> e.g - consider you were saving a date time value as `string`, and now you altered the data type to `date_time` in new version. When being converted to datatype, there might be that there is no value in the column or value is not date convertable.
> 
> That's why i decided that user should convert the value manually as it provides command over your code & data.
>
You are totally right on that. Maybe for just simple convertions number=> string or string => number when all the field have already all numbers, but i dont care about that.

Are you familiare with the android sqliteDatabase solution ? In there class they have also an version control. When there is a version number change an onUpgade event is called with the old and new version. When a new user is starting the app it using the db an onCreate event is executed to allow creating the database.


Is this something you can do? 
Or do i just don't understand how you implemented it it all together and i need something that is not in possible :(
Could you maybe explain? Are you need the DbSchema all the time to get the data back from indexdb? I'm totally new to this db implementation. sorry.

**Comment by ujjwalguptaofficial** [comment]
No worry, i can understand it can be little complex. You will understand it after some time.

> events i can work on but i don't see any need. Anyway will work on this. 

But for you to understand, here is example POC - https://github.com/ujjwalguptaofficial/jsstore-examples/tree/master/change-db-schema

 Run the example and fill some data and then click on convert to v2. If unable to understand, just debug the code line by line.
 
 In the example at line - https://github.com/ujjwalguptaofficial/jsstore-examples/blob/be79f5276af2fbc3bfc6da7cd97738aa0caa80a9/change-db-schema/scripts/index.js#L17   you can check how i am converting from one data type to another.

**Comment by nielsnl68** [comment]
Ha yes, it getting somewhere.
Just to be clear, the column section is always version 1 and should not be changed after that, all changes to the schema should be added in the in the alter section.
How about an user starts with version 1 ... does not use it for some time, and when he use it again the version is 5. 
Will jsStore get the change from 2..5 from the alter section? And what happens when 3 is just an update to the values and the schema stay's the same?  Will initDB() still give true back?

I have been thinking an event is maybe overkill, just an callback function like:

``` javascript
initDB(dbSchema, dbupgrade_cb);

// where dbupgrade_cb looks like below, note there are no break; betweens cases, 
// that is intentionally:

function dbupgrade_cb(db, oldVersion, newVersion) {
    switch (oldVersion) {
        case 0:
            /// initiate the db on with some data
        case 1:
            var datas = await jsstoreCon.select({ from: "Student" });
            datas.forEach(data => {
                const gender = data.gender;
                if (!gender) {
                    data.gender = 0;
                } else {
                    data.gender = gender === "male" ? 1 : 2;
                }
            })
            await jsstoreCon.insert({ into: "Student", values: datas, upsert: true })
        case 2:
        // do something here 
        case 3:
        // something else
    }
}
```

**Comment by ujjwalguptaofficial** [comment]
> How about an user starts with version 1 ... does not use it for some time, and when he use it again the version is 5.

nice thought, this is not handled. I will work right away.

> And what happens when 3 is just an update to the values and the schema stay's the same? Will initDB() still give true back?

yep initDb will return true whenever version is changed or database is created for first time. 

> I have been thinking an event is maybe overkill, just an callback function like:

I also think, callback might be a good idea.

**Comment by ujjwalguptaofficial** [comment]
fixed all issue in v - 4.2.2.

So now 

* all alter scripts starting from oldversion to newversion will be executed. 
* events will be emiitted for open, upgrade, create.  Read about events here - https://jsstore.net/tutorial/event/
